packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -454,127 +461,296 @@ export class ModularChangeFamily
 		}
 		const revisionMetadata: RevisionMetadataSource = revisionMetadataSourceFromInfo(revInfos);
 		const rebasedFields = this.rebaseFieldMap(
-			change.changes,
-			tagChange(over.change.changes, over.revision),
+			change.fieldChanges,
+			tagChange(over.change.fieldChanges, over.revision),
 			genId,
 			crossFieldTable,
+			() => true,
 			revisionMetadata,
 		);
 
-		if (crossFieldTable.fieldsToUpdate.size > 0) {
-			const fieldsToUpdate = crossFieldTable.fieldsToUpdate;
-			crossFieldTable.fieldsToUpdate = new Set();
-			for (const { fieldChange, baseChange } of fieldsToUpdate) {
-				const amendedChange = getChangeHandler(
-					this.fieldKinds,
-					fieldChange.fieldKind,
-				).rebaser.amendRebase(
-					fieldChange.change,
-					baseChange,
-					genId,
-					newCrossFieldManager(crossFieldTable),
-					revisionMetadata,
-				);
-				fieldChange.change = brand(amendedChange);
-			}
-		}
+		const rebasedChangeset = makeModularChangeset(rebasedFields, maxId, change.revisions);
+		crossFieldTable.baseMapToRebased.set(over.change.fieldChanges, rebasedChangeset);
+
+		const [fieldsToAmend, invalidatedEmptyFields] = getFieldsToAmend(
+			crossFieldTable.invalidatedFields,
+			crossFieldTable.baseMapToRebased,
+			crossFieldTable.baseChangeToContext,
+			crossFieldTable.baseMapToParentField,
+		);
 
+		this.amendRebase(
+			crossFieldTable,
+			fieldsToAmend,
+			invalidatedEmptyFields,
+			genId,
+			revisionMetadata,
+		);
 		assert(
-			crossFieldTable.fieldsToUpdate.size === 0,
+			crossFieldTable.invalidatedFields.size === 0,
 			0x59f /* Should not need more than one amend pass. */,
 		);
 
-		return makeModularChangeset(rebasedFields, maxId, change.revisions);
+		if (maxId >= 0) {
+			rebasedChangeset.maxId = brand(maxId);
+		}
+		return rebasedChangeset;
+	}
+
+	private amendRebase(
+		crossFieldTable: RebaseTable,
+		fieldsToAmend: Set<FieldChange>,
+		invalidatedEmptyFields: Set<FieldChange>,
+		genId: IdAllocator,
+		revisionMetadata: RevisionMetadataSource,
+	) {
+		crossFieldTable.invalidatedFields = new Set();
+		for (const fieldToAmend of fieldsToAmend) {
+			const fieldContext = crossFieldTable.baseChangeToContext.get(fieldToAmend);
+			let fieldChange: FieldChange;
```

Comment:
Nit: consider renaming `fieldChange` to `newChange` or `newFieldChange`. The current name discribes the type, which most tooling already helps with, but does not describe which of the change changesets the `FieldChange` belongs to, which is the more relevant question here.